src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+        """
+        Build common-lines matrix from Fourier stack of 2D images
+
+        :param n_check: For each image/projection find its common-lines with
+            n_check images. If n_check is less than the total number of images,
+            a random subset of n_check images is used.
+        """
+
+        n_img = self.n_img
+        if n_check is None:
+            n_check = n_img
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+
+        # need to do a copy to prevent modifying self.pf for other functions
+        pf = self.pf.copy()
+
+        # Allocate local variables for return
+        # clstack represents the common lines matrix:
+        # clstack[i, j] and clstack[j ,i] contain the index
+        # of the common line of projections i and j. Namely,
+        # clstack[i,j] contains the index of the common line
+        # in the image i. clstack[j,i] contains the index of
+        # the common line in j image.
+        clstack = -np.ones((n_img, n_img))
+        # corrstack defines the correlation of the common line
+        # between image i and j. Since corrstack is symmetric,
+        # only above the diagonal entries are necessary.
+        # corrstack[i, j] measures how ''common'' is the between
+        # image i and j. Small value means high-similarity.
+        corrstack = np.zeros((n_img, n_img))
+
+        # Allocate variables used for shift estimation
+
+        # set maximum value of 1D shift (in pixels) to search
+        # between common-lines.
+        max_shift = self.max_shift
+        # Set resolution of shift estimation in pixels. Note that
+        # shift_step can be any positive real number.
+        shift_step = self.shift_step
+        # 1D shift between common-lines
+        shifts_1d = np.zeros((n_img, n_img))
+
+        # Prepare the shift phases to try and generate filter for common-line detection
+        r_max = int((pf.shape[0] - 1) / 2)
+        shifts, shift_phases, h = self._generate_shift_phase_and_filter(r_max, max_shift, shift_step)
+        all_shift_phases = shift_phases.T
+
+        # Apply bandpass filter and normalize each ray of each image.
+        np.einsum('ijk, i -> ijk', pf, h, out=pf)
+        pf[r_max - 1:r_max + 2] = 0
+        pf /= np.linalg.norm(pf, axis=0)
+
+        # Only half of each ray is needed
+        pf = pf[0:r_max]
+        pf = pf.transpose((2, 1, 0))
+
+        # Search for common lines between [i, j] pairs of images.
+        # Creating pf and building common lines are different to the Matlab version.
+        # The random selection is implemented.
+        for i in range(n_img-1):
+
+            p1 = pf[i]
+            p1_real = np.real(p1)
+            p1_imag = np.imag(p1)
+
+            # build the subset of j images if n_check < n_img
+            n2 = min(n_img - i, n_check)
+            subset_j = np.sort(np.random.choice(n_img - i - 1, n2 - 1,
```

Comment:
Yes. This is because of `n_img- i -1` in front and `i + 1` in the next line, and the result of `random.choice` could be 0. Any way, the `subset_j` should be in the range of `i + 1` to `n_img` at most.

Reply:
I don't understand. If we're supposed to compare with `n_check` images, why are we drawing `n_check - 1` here? It looks more likely that the `n_img - i` in the line above should be replaced by `n_img - i - 1`, then we can just have `n2` on this line. Also, we need a better name than `n2`, which doesn't mean anything.

Reply:
The j value starts from i+1 and we have excluded j=i. So the rest total number to choose is n_check-1 or n_img - 1 (for n_check = n_img). Changed n2 to num_j, the number of j values.

Reply:
> So the rest total number to choose is n_check-1 or n_img - 1 (for n_check = n_img). 

I understand that. But if `n_check` is *strictly less* than the number of images, we only end up checking `n_check - 1` of them, which is confusing. The code should read
```
n_remaining = n_img - i - 1
n_j = min(n_remaining,  n_check)
subset_j = np.sort(n_remaining, n_j, replace=False) + i + 1)
```
It's also not clear that we need `np.sort` here.

Reply:
Done. The sort is need to generate the same order such as j can go from `i + 1` to `n_img -1` for `n_check = n_img`.